In [1]:
import polars as pl
import numpy as np

### Paso 1. Carga de Datos y Visualización

#### Carga de datos y Visualización del Tamaño de los Datos

In [2]:
# Cargar datos preparados tras el proceso de limpieza y EDA
df = pl.read_parquet("../data/modeling/base_modelado.parquet")

print(f"Filas: {df.height:,}")
print(f"Columnas: {df.width}")

df.head()

Filas: 40,521,392
Columnas: 24


id_sensor,id_fecha,hora,intensidad_media,intensidad_max,intensidad_min,ocupacion_media,ocupacion_max,velocidad_media,velocidad_min,num_mediciones,num_error_E,porcentaje_calidad,año,mes,nombre_mes,trimestre,dia,dia_semana,fin_semana,tipo_elem,distrito,latitud,longitud
i32,date,i32,f64,f64,f64,f64,f64,f64,f64,i64,f64,f64,i64,i64,str,i64,i64,i64,bool,str,i32,f64,f64
1001,2026-04-18,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,4,"""April""",2,18,6,true,"""other""",10,40.409729,-3.740786
1001,2026-04-21,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,4,"""April""",2,21,2,false,"""other""",10,40.409729,-3.740786
1001,2026-04-21,6,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,4,"""April""",2,21,2,false,"""other""",10,40.409729,-3.740786
1002,2026-04-16,10,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,4,"""April""",2,16,4,false,"""other""",10,40.40803,-3.74376
1002,2026-04-17,21,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,4,"""April""",2,17,5,false,"""other""",10,40.40803,-3.74376


#### Visualización de la Estructura Interna de los Datos

In [3]:
df.schema

Schema([('id_sensor', Int32),
        ('id_fecha', Date),
        ('hora', Int32),
        ('intensidad_media', Float64),
        ('intensidad_max', Float64),
        ('intensidad_min', Float64),
        ('ocupacion_media', Float64),
        ('ocupacion_max', Float64),
        ('velocidad_media', Float64),
        ('velocidad_min', Float64),
        ('num_mediciones', Int64),
        ('num_error_E', Float64),
        ('porcentaje_calidad', Float64),
        ('año', Int64),
        ('mes', Int64),
        ('nombre_mes', String),
        ('trimestre', Int64),
        ('dia', Int64),
        ('dia_semana', Int64),
        ('fin_semana', Boolean),
        ('tipo_elem', String),
        ('distrito', Int32),
        ('latitud', Float64),
        ('longitud', Float64)])

#### Visualización de la Dsitribución de los Datos

In [4]:
df.describe()

statistic,id_sensor,id_fecha,hora,intensidad_media,intensidad_max,intensidad_min,ocupacion_media,ocupacion_max,velocidad_media,velocidad_min,num_mediciones,num_error_E,porcentaje_calidad,año,mes,nombre_mes,trimestre,dia,dia_semana,fin_semana,tipo_elem,distrito,latitud,longitud
str,f64,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,str,f64,f64,f64,f64,str,f64,f64,f64
"""count""",4.0521392e7,"""40521392""",4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,"""40521392""",4.0521392e7,4.0521392e7,4.0521392e7,4.0521392e7,"""40521392""",4.0521392e7,4.0521392e7,4.0521392e7
"""null_count""",0.0,"""0""",0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,"""0""",0.0,0.0,0.0,0.0,"""0""",0.0,0.0,0.0
"""mean""",6794.155899,"""2025-12-30 16:06:56.105902""",11.520612,334.255552,383.585398,287.305949,5.236373,7.375898,5.116786,4.755918,3.960639,0.0,99.517551,2025.499504,6.503937,null,2.50404,15.747258,2.999349,0.281935,null,10.287664,40.430422,-3.684039
"""std""",2677.698222,null,6.917672,576.100512,629.976612,528.555676,9.319953,11.967622,18.415437,17.488734,0.278023,0.0,6.845825,0.5,3.448665,null,1.115256,8.765745,1.992627,null,null,5.38041,0.038802,0.042648
"""min""",1001.0,"""2025-07-01""",0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,2025.0,1.0,"""April""",1.0,1.0,0.0,0.0,"""M30""",0.0,40.332454,-3.836886
"""25%""",4693.0,"""2025-09-30""",6.0,47.5,63.0,32.0,0.5,1.0,0.0,0.0,4.0,0.0,100.0,2025.0,4.0,null,2.0,8.0,1.0,null,null,6.0,40.39921,-3.712531
"""50%""",5989.0,"""2025-12-31""",12.0,150.5,184.0,118.0,2.25,3.0,0.0,0.0,4.0,0.0,100.0,2025.0,7.0,null,3.0,16.0,3.0,null,null,10.0,40.43189,-3.68692
"""75%""",10111.0,"""2026-04-02""",18.0,387.5,450.0,328.0,5.75,8.0,0.0,0.0,4.0,0.0,100.0,2026.0,10.0,null,4.0,23.0,5.0,null,null,15.0,40.459606,-3.656483
"""max""",11541.0,"""2026-06-30""",23.0,91368.333333,93520.0,91034.0,100.0,100.0,246.0,246.0,4.0,0.0,100.0,2026.0,12.0,"""September""",4.0,31.0,6.0,1.0,"""other""",21.0,40.515611,-3.551623


### Paso 2: Transformaciones

#### Creación de una columna de tipo fecha y hora a partir del id

In [5]:
df = (
    df.with_columns(
        pl.datetime(
            year=pl.col("id_fecha").dt.year(),
            month=pl.col("id_fecha").dt.month(),
            day=pl.col("id_fecha").dt.day(),
            hour=pl.col("hora"),
            minute=pl.lit(0),
            second=pl.lit(0),
        ).alias("fecha_hora")
    )
    .sort(["id_sensor", "fecha_hora"])
)

#### Visualización de las columnas de fecha generadas

In [6]:
df.select(
    [
        "id_sensor",
        "id_fecha",
        "hora",
        "fecha_hora",
    ]
).head(10)

id_sensor,id_fecha,hora,fecha_hora
i32,date,i32,datetime[μs]
1001,2026-03-04,13,2026-03-04 13:00:00
1001,2026-03-04,14,2026-03-04 14:00:00
1001,2026-03-04,15,2026-03-04 15:00:00
1001,2026-03-04,16,2026-03-04 16:00:00
1001,2026-03-04,17,2026-03-04 17:00:00
1001,2026-03-04,18,2026-03-04 18:00:00
1001,2026-03-04,19,2026-03-04 19:00:00
1001,2026-03-04,20,2026-03-04 20:00:00
1001,2026-03-04,21,2026-03-04 21:00:00


#### Visualización de la Agrupación por Sensor

In [7]:
(
    df
    .group_by("id_sensor")
    .agg(
        [
            pl.col("fecha_hora").min().alias("inicio"),
            pl.col("fecha_hora").max().alias("fin"),
            pl.len().alias("registros"),
        ]
    )
    .head()
)

id_sensor,inicio,fin,registros
i32,datetime[μs],datetime[μs],u32
1001,2026-03-04 13:00:00,2026-06-17 01:00:00,1854
1002,2026-03-04 13:00:00,2026-06-17 01:00:00,1854
1003,2026-03-04 13:00:00,2026-06-17 01:00:00,1854
1006,2025-07-01 00:00:00,2026-06-30 23:00:00,7979
1009,2025-09-25 09:00:00,2026-05-26 13:00:00,5081


#### Calculo dse del tiempo transcurrido entre una medición y la anterior

In [8]:
df = df.with_columns(
    pl.col("fecha_hora")
    .diff()
    .over("id_sensor")
    .alias("delta_tiempo")
)

#### Visualización de la frecuencia de los intervalos de tiempo

In [9]:
(
    df
    .group_by("delta_tiempo")
    .len()
    .sort("delta_tiempo")
)

delta_tiempo,len
duration[μs],u32
null,4933
1h,40384131
2h,56185
3h,21554
4h,9218
…,…
169d 5h,1
171d 17h,1
173d 12h,1


#### Detección de todos los registros en los que existe un salto temporal entre mediciones

In [10]:
huecos = (
    df
    .filter(
        (pl.col("delta_tiempo").is_not_null()) &
        (pl.col("delta_tiempo") != pl.duration(hours=1))
    )
)

print(f"Registros con saltos temporales: {huecos.height:,}")

Registros con saltos temporales: 132,328


#### Visualización de los huecos detectados

In [11]:
huecos.select(
    [
        "id_sensor",
        "fecha_hora",
        "delta_tiempo",
    ]
).head(20)

id_sensor,fecha_hora,delta_tiempo
i32,datetime[μs],duration[μs]
1001,2026-04-16 09:00:00,27d 8h
1002,2026-04-16 09:00:00,27d 8h
1003,2026-04-16 09:00:00,27d 8h
1006,2025-07-24 07:00:00,4d 22h
1006,2025-07-25 05:00:00,4h
…,…,…
1006,2025-09-26 08:00:00,7h
1006,2025-10-10 03:00:00,2h
1006,2025-11-12 20:00:00,6h


#### Eliminación de la columna `delta_tiempo`

In [12]:
df = df.drop("delta_tiempo")

#### Visualización de valores duplicados

In [13]:
duplicados = (
    df
    .group_by(["id_sensor", "fecha_hora"])
    .len()
    .filter(pl.col("len") > 1)
)

print(f"Duplicados encontrados: {duplicados.height}")

Duplicados encontrados: 0


In [14]:
print("=" * 50)
print("PREPARACIÓN COMPLETADA")
print("=" * 50)

print(f"Filas: {df.height:,}")
print(f"Columnas: {df.width:,}")
print(f"Sensores únicos: {df['id_sensor'].n_unique():,}")

PREPARACIÓN COMPLETADA
Filas: 40,521,392
Columnas: 25
Sensores únicos: 4,933


### Paso 3: Feature Engineering

#### Transformación de las variables temporales en variables cíclicas usando las funciones seno y coseno

In [15]:
# Transformación cíclica de variables temporales

df = df.with_columns([
    (2 * np.pi * pl.col("hora") / 24).sin().alias("hora_sin"),
    (2 * np.pi * pl.col("hora") / 24).cos().alias("hora_cos"),

    (2 * np.pi * pl.col("dia_semana") / 7).sin().alias("dia_semana_sin"),
    (2 * np.pi * pl.col("dia_semana") / 7).cos().alias("dia_semana_cos"),

    (2 * np.pi * pl.col("mes") / 12).sin().alias("mes_sin"),
    (2 * np.pi * pl.col("mes") / 12).cos().alias("mes_cos"),
])

#### Visualización de las tranformaciones realizadas

In [16]:
df.select([
    "hora",
    "hora_sin",
    "hora_cos"
]).head()

hora,hora_sin,hora_cos
i32,f64,f64
13,-0.258819,-0.965926
14,-0.5,-0.866025
15,-0.707107,-0.707107
16,-0.866025,-0.5
17,-0.965926,-0.258819


#### definición de las variables del dataset que van a utilizarse para crear lag features

In [17]:
variables_lag = [
    "intensidad_media",
    "ocupacion_media",
    "velocidad_media"
]

lags = [1, 24, 168]

#### Creación de las lag_features

In [18]:
lag_exprs = []

for variable in variables_lag:
    for lag in lags:
        lag_exprs.append(
            pl.col(variable)
            .shift(lag)
            .over("id_sensor")
            .alias(f"{variable}_lag_{lag}")
        )

df = df.with_columns(lag_exprs)

#### Visualización de las lag_features

In [19]:
df.select([
    "ocupacion_media",
    "ocupacion_media_lag_1",
    "ocupacion_media_lag_24",
    "ocupacion_media_lag_168"
]).head(10)

ocupacion_media,ocupacion_media_lag_1,ocupacion_media_lag_24,ocupacion_media_lag_168
f64,f64,f64,f64
0.0,null,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null


#### Creación de las columnas con estadísticas móviles

In [20]:
rolling_exprs = [

    # Intensidad
    pl.col("intensidad_media")
    .rolling_mean(window_size=3)
    .over("id_sensor")
    .alias("rolling_intensidad_mean_3"),

    pl.col("intensidad_media")
    .rolling_mean(window_size=24)
    .over("id_sensor")
    .alias("rolling_intensidad_mean_24"),

    pl.col("intensidad_media")
    .rolling_std(window_size=24)
    .over("id_sensor")
    .alias("rolling_intensidad_std_24"),


    # Ocupación
    pl.col("ocupacion_media")
    .rolling_mean(window_size=3)
    .over("id_sensor")
    .alias("rolling_ocupacion_mean_3"),

    pl.col("ocupacion_media")
    .rolling_mean(window_size=24)
    .over("id_sensor")
    .alias("rolling_ocupacion_mean_24"),

    pl.col("ocupacion_media")
    .rolling_std(window_size=24)
    .over("id_sensor")
    .alias("rolling_ocupacion_std_24"),

]

df = df.with_columns(rolling_exprs)

#### Visualización de las variables creadas sobre las estadísticas móviles

In [21]:
df.select([
    "ocupacion_media",
    "rolling_ocupacion_mean_3",
    "rolling_ocupacion_mean_24",
    "rolling_ocupacion_std_24"
]).head(10)

ocupacion_media,rolling_ocupacion_mean_3,rolling_ocupacion_mean_24,rolling_ocupacion_std_24
f64,f64,f64,f64
0.0,null,null,null
0.0,null,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null
0.0,0.0,null,null


#### Cálculo de diferencias puntuales respecto a períodos anteriores

In [22]:
df = df.with_columns([

    # Intensidad
    (
        pl.col("intensidad_media")
        - pl.col("intensidad_media_lag_1")
    ).alias("delta_intensidad_1"),

    (
        pl.col("intensidad_media")
        - pl.col("intensidad_media_lag_24")
    ).alias("delta_intensidad_24"),

    # Ocupación
    (
        pl.col("ocupacion_media")
        - pl.col("ocupacion_media_lag_1")
    ).alias("delta_ocupacion_1"),

    (
        pl.col("ocupacion_media")
        - pl.col("ocupacion_media_lag_24")
    ).alias("delta_ocupacion_24"),

])

#### Visualización de las variables creadas anteriormente

In [23]:
df.select([
    "ocupacion_media",
    "ocupacion_media_lag_1",
    "delta_ocupacion_1",
    "ocupacion_media_lag_24",
    "delta_ocupacion_24"
]).head(10)

ocupacion_media,ocupacion_media_lag_1,delta_ocupacion_1,ocupacion_media_lag_24,delta_ocupacion_24
f64,f64,f64,f64,f64
0.0,null,null,null,null
0.0,0.0,0.0,null,null
0.0,0.0,0.0,null,null
0.0,0.0,0.0,null,null
0.0,0.0,0.0,null,null
0.0,0.0,0.0,null,null
0.0,0.0,0.0,null,null
0.0,0.0,0.0,null,null
0.0,0.0,0.0,null,null


#### Creación de la variable target para el entrenamiento del modelo de Machine Learning

In [24]:
df = df.with_columns(
    pl.col("ocupacion_media")
    .shift(-1)
    .over("id_sensor")
    .alias("target")
)

#### Visualizaci´n de la variable target

In [25]:
df.select([
    "fecha_hora",
    "ocupacion_media",
    "target"
]).head(10)

fecha_hora,ocupacion_media,target
datetime[μs],f64,f64
2026-03-04 13:00:00,0.0,0.0
2026-03-04 14:00:00,0.0,0.0
2026-03-04 15:00:00,0.0,0.0
2026-03-04 16:00:00,0.0,0.0
2026-03-04 17:00:00,0.0,0.0
2026-03-04 18:00:00,0.0,0.0
2026-03-04 19:00:00,0.0,0.0
2026-03-04 20:00:00,0.0,0.0
2026-03-04 21:00:00,0.0,0.0


#### Filtrado de targets inválidos por saltos temporales

El `shift(-1)` asume que la fila siguiente de cada sensor corresponde exactamente a la hora siguiente. Sin embargo, existen 132.328 saltos temporales (huecos de 2h a 173 días). En esas filas el target no es la ocupación de t+1 sino la de un instante muy posterior — etiquetas incorrectas que contaminan el entrenamiento. Se filtran las filas cuyo registro siguiente no está exactamente a 1 hora.

In [26]:
filas_antes_filtro = df.height

df = df.with_columns(
    pl.col("fecha_hora")
    .shift(-1)
    .over("id_sensor")
    .alias("fecha_hora_siguiente")
)

df = df.filter(
    (pl.col("fecha_hora_siguiente") - pl.col("fecha_hora"))
    == pl.duration(hours=1)
).drop("fecha_hora_siguiente")

filas_despues_filtro = df.height

print(f"Filas antes del filtro : {filas_antes_filtro:,}")
print(f"Filas después del filtro: {filas_despues_filtro:,}")
print(f"Targets inválidos eliminados: {filas_antes_filtro - filas_despues_filtro:,}")

Filas antes del filtro : 40,521,392
Filas después del filtro: 40,384,131
Targets inválidos eliminados: 137,261


In [27]:
(
    df
    .null_count()
    .transpose(include_header=True)
    .filter(pl.col("column_0") > 0)
)

column,column_0
str,u32
"""intensidad_media_lag_1""",4825
"""intensidad_media_lag_24""",117044
"""intensidad_media_lag_168""",821709
"""ocupacion_media_lag_1""",4825
"""ocupacion_media_lag_24""",117044
…,…
"""rolling_ocupacion_std_24""",112158
"""delta_intensidad_1""",4825
"""delta_intensidad_24""",117044


In [28]:
filas_antes = df.height

df = df.drop_nulls()

filas_despues = df.height

print(f"Filas iniciales: {filas_antes:,}")
print(f"Filas finales: {filas_despues:,}")
print(f"Filas eliminadas: {filas_antes - filas_despues:,}")

Filas iniciales: 40,384,131
Filas finales: 39,562,422
Filas eliminadas: 821,709


In [29]:
df = df.drop([
    "id_fecha",
    "nombre_mes"
])

In [30]:
print(df.shape)

(39562422, 49)


In [31]:
df.head()

id_sensor,hora,intensidad_media,intensidad_max,intensidad_min,ocupacion_media,ocupacion_max,velocidad_media,velocidad_min,num_mediciones,num_error_E,porcentaje_calidad,año,mes,trimestre,dia,dia_semana,fin_semana,tipo_elem,distrito,latitud,longitud,fecha_hora,hora_sin,hora_cos,dia_semana_sin,dia_semana_cos,mes_sin,mes_cos,intensidad_media_lag_1,intensidad_media_lag_24,intensidad_media_lag_168,ocupacion_media_lag_1,ocupacion_media_lag_24,ocupacion_media_lag_168,velocidad_media_lag_1,velocidad_media_lag_24,velocidad_media_lag_168,rolling_intensidad_mean_3,rolling_intensidad_mean_24,rolling_intensidad_std_24,rolling_ocupacion_mean_3,rolling_ocupacion_mean_24,rolling_ocupacion_std_24,delta_intensidad_1,delta_intensidad_24,delta_ocupacion_1,delta_ocupacion_24,target
i32,i32,f64,f64,f64,f64,f64,f64,f64,i64,f64,f64,i64,i64,i64,i64,i64,bool,str,i32,f64,f64,datetime[μs],f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
1001,13,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,3,1,11,3,false,"""other""",10,40.409729,-3.740786,2026-03-11 13:00:00,-0.258819,-0.965926,0.433884,-0.900969,1.0,6.1232e-17,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1001,14,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,3,1,11,3,false,"""other""",10,40.409729,-3.740786,2026-03-11 14:00:00,-0.5,-0.866025,0.433884,-0.900969,1.0,6.1232e-17,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1001,15,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,3,1,11,3,false,"""other""",10,40.409729,-3.740786,2026-03-11 15:00:00,-0.707107,-0.707107,0.433884,-0.900969,1.0,6.1232e-17,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1001,16,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,3,1,11,3,false,"""other""",10,40.409729,-3.740786,2026-03-11 16:00:00,-0.866025,-0.5,0.433884,-0.900969,1.0,6.1232e-17,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1001,17,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4,0.0,100.0,2026,3,1,11,3,false,"""other""",10,40.409729,-3.740786,2026-03-11 17:00:00,-0.965926,-0.258819,0.433884,-0.900969,1.0,6.1232e-17,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [32]:
df.describe()

statistic,id_sensor,hora,intensidad_media,intensidad_max,intensidad_min,ocupacion_media,ocupacion_max,velocidad_media,velocidad_min,num_mediciones,num_error_E,porcentaje_calidad,año,mes,trimestre,dia,dia_semana,fin_semana,tipo_elem,distrito,latitud,longitud,fecha_hora,hora_sin,hora_cos,dia_semana_sin,dia_semana_cos,mes_sin,mes_cos,intensidad_media_lag_1,intensidad_media_lag_24,intensidad_media_lag_168,ocupacion_media_lag_1,ocupacion_media_lag_24,ocupacion_media_lag_168,velocidad_media_lag_1,velocidad_media_lag_24,velocidad_media_lag_168,rolling_intensidad_mean_3,rolling_intensidad_mean_24,rolling_intensidad_std_24,rolling_ocupacion_mean_3,rolling_ocupacion_mean_24,rolling_ocupacion_std_24,delta_intensidad_1,delta_intensidad_24,delta_ocupacion_1,delta_ocupacion_24,target
str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,str,f64,f64,f64,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""count""",3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,"""39562422""",3.9562422e7,3.9562422e7,3.9562422e7,"""39562422""",3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7,3.9562422e7
"""null_count""",0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,"""0""",0.0,0.0,0.0,"""0""",0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
"""mean""",6791.556705,11.522935,334.388518,383.77,287.377021,5.226758,7.370109,5.110759,4.749143,3.966344,0.0,99.515134,2025.509267,6.494262,2.494376,15.975322,2.998974,0.281956,null,10.291601,40.430376,-3.68402,"""2026-01-03 15:54:20.064785""",-0.001964,-0.002703,0.00368,-0.004216,0.006888,0.012617,334.455145,334.438768,334.093789,5.227416,5.229038,5.223735,5.110438,5.110786,5.111706,334.422447,334.318999,193.3432,5.227025,5.225789,3.893396,-0.066628,-0.05025,-0.000658,-0.00228,5.226673
"""std""",2676.055286,6.912909,576.226043,629.943621,528.699694,9.266514,11.930043,18.403987,17.475067,0.252722,0.0,6.862339,0.499914,3.481256,1.123887,8.686801,1.992735,null,null,5.381476,0.038783,0.042649,null,0.706999,0.707207,0.708796,0.705391,0.707535,0.706533,576.58053,576.673109,576.283142,9.271774,9.282285,9.288025,18.403035,18.40422,18.405835,565.114897,481.027443,260.624955,8.850886,7.266071,4.422234,158.959627,219.355106,4.197514,5.165772,9.270017
"""min""",1001.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,2025.0,1.0,1.0,1.0,0.0,0.0,"""M30""",0.0,40.332454,-3.836886,"""2025-07-08 00:00:00""",-1.0,-1.0,-0.974928,-0.900969,-1.0,-1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,-1.2127e-12,-9.4739e-15,0.0,-2.8422e-14,-5.9212e-16,0.0,-87637.0,-89937.25,-100.0,-100.0,0.0
"""25%""",4693.0,6.0,47.75,64.0,32.0,0.5,1.0,0.0,0.0,4.0,0.0,100.0,2025.0,3.0,1.0,9.0,1.0,null,null,6.0,40.399208,-3.712531,"""2025-10-06 03:00:00""",-0.707107,-0.707107,-0.781831,-0.900969,-0.866025,-0.5,47.75,47.75,47.5,0.5,0.5,0.5,0.0,0.0,0.0,52.0,87.15625,52.259057,0.666667,1.479167,1.182938,-29.25,-21.25,-0.75,-0.75,0.5
"""50%""",5988.0,12.0,151.0,184.0,118.0,2.25,3.0,0.0,0.0,4.0,0.0,100.0,2026.0,6.0,2.0,16.0,3.0,null,null,10.0,40.431868,-3.686914,"""2026-01-04 05:00:00""",0.0,-1.8370e-16,0.0,-0.222521,1.2246e-16,6.1232e-17,151.0,150.75,150.5,2.25,2.25,2.25,0.0,0.0,0.0,155.666667,190.875,113.306037,2.333333,3.072917,2.312439,-1.25,0.0,0.0,0.0,2.25
"""75%""",10108.0,18.0,388.0,451.0,328.0,5.75,8.0,0.0,0.0,4.0,0.0,100.0,2026.0,10.0,4.0,23.0,5.0,null,null,15.0,40.459552,-3.65645,"""2026-04-03 17:00:00""",0.707107,0.707107,0.781831,0.62349,0.5,0.866025,388.0,387.75,387.25,5.75,5.75,5.75,0.0,0.0,0.0,389.7

In [33]:
print("=" * 60)
print("FEATURE ENGINEERING COMPLETADO")
print("=" * 60)

print(f"Filas finales: {df.height:,}")
print(f"Columnas finales: {df.width}")
print(f"Sensores únicos: {df['id_sensor'].n_unique():,}")

print("\nEl dataset incluye:")
print("- Variables temporales originales")
print("- Variables temporales cíclicas")
print("- Variables lag")
print("- Variables rolling")
print("- Variables de tendencia")
print("- Variable objetivo (target)")

FEATURE ENGINEERING COMPLETADO
Filas finales: 39,562,422
Columnas finales: 49
Sensores únicos: 4,915

El dataset incluye:
- Variables temporales originales
- Variables temporales cíclicas
- Variables lag
- Variables rolling
- Variables de tendencia
- Variable objetivo (target)


### Paso 4: Almacenado de los datos procesados

In [34]:
df.write_parquet("../data/modeling/features_v2.parquet")

print("Feature engineering completado correctamente.")

Feature engineering completado correctamente.
